File: LCOF 46/translateNum.py
class Solution:
    def translateNum(self, num: int) -> int:
        strNum = str(num)
        strLen = len(strNum)
        if strLen == 1:
            return 1

        dp = [1] * (strLen + 1)
        for i in range(1, strLen):
            preSum = int(strNum[i - 1]) * 10 + int(strNum[i])
            if preSum < 26 and preSum != int(strNum[i]):
                dp[i + 1] = dp[i] + dp[i - 1]
            else:
                dp[i + 1] = dp[i]

        return dp[strLen]

File: LCOF 46/translateNum.java
class Solution {
    public int translateNum(int num) {
        String strNum = String.valueOf(num);
        int len = strNum.length();
        if (len == 1) {
            return 1;
        }

        int[] dp = new int[len + 1];
        dp[0] = 1;
        dp[1] = 1;
        for (int i = 1; i < len; i++) {
            int preSum = (strNum.charAt(i - 1) - '0') * 10 + (strNum.charAt(i) - '0');
            if (preSum < 26 && preSum != (strNum.charAt(i) - '0')) {
                dp[i + 1] = dp[i] + dp[i - 1];
            } else {
                dp[i + 1] = dp[i];
            }
        }
        return dp[len];
    }
}